include/envoy/http/codec.h

```diff
@@ -73,34 +53,57 @@ class StreamEncoder {
 };
 
 /**
- * Stream encoder used for making a request (client to server).
+ * Stream encoder used for sending a request (client to server).
  */
-class RequestStreamEncoder : public virtual StreamEncoder {};
+class RequestEncoder : public virtual StreamEncoder {
+public:
+  /**
+   * Encode headers, optionally indicating end of stream. Response headers must
+   * have a valid :status set.
+   * @param headers supplies the header map to encode.
+   * @param end_stream supplies whether this is a header only request/response.
+   */
+  virtual void encodeHeaders(const HeaderMap& headers, bool end_stream) PURE;
```

Comment:
I'm ambivalent on this. If we move the other methods, I don't see the point on having the base StreamEncoder class at all, and getStream() would be duplicated in both. Similar for StreamDecoder. It will be a bunch of work. My preference would be to skip this for now as IMO this could be done as a future cleanup and is somewhat orthogonal to my current goals, but if you prefer I can do it now.